### PR TITLE
[batch/job] Partial admission - sync completions

### PR DESCRIPTION
**Pull request by trasc**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
Synchronize `batch/job.completions` with `parallelism` in case of partial admission

#### Which issue(s) this PR fixes:

Relates to: #970

#### Special notes for your reviewer:

This is the first step fixing #970.

#### Does this PR introduce a user-facing change?

```release-note
Synchronize batch/job.completions with parallelism in case of partial admission
```

### CODE REVIEW

File: pkg/controller/jobs/job/job_webhook.go [hunk 1]
```diff
@@ -117,6 +124,15 @@ func validatePartialAdmissionCreate(job *Job) field.ErrorList {
...
 	}
+	if strVal, found := job.Annotations[JobSyncCompletionWithParallelism]; found {
+		enabled, err := strconv.ParseBool(strVal)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(syncCompletionAnnotationsPath, job.Annotations[JobSyncCompletionWithParallelism], err.Error()))
+		}
+		if enabled && pointer.Int32Deref(job.Spec.Parallelism, 1) != pointer.Int32Deref(job.Spec.Completions, 1) {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "completions"), job.Spec.Completions, "should be equal to parallelism when sync completions is enabled"))
+		}
+	}
```
Comment (tenzen-y) [inline pkg/controller/jobs/job/job_webhook.go:135]: Should we add a validation to reject jobs with `NonIndexed` mode and `kueue.x-k8s.io/job-completions-equal-parallelism=true` annotation?
IIUC, the `completions` field is immutable when the job is created with `NonIndexed` mode.

**Review by tenzen-y — COMMENTED** [review]
Generally, LGTM.

File: pkg/controller/jobs/job/job_webhook.go (hunk 1 above)
Comment (alculquicondor) [reply pkg/controller/jobs/job/job_webhook.go:135]: +1

Comment (trasc) [reply pkg/controller/jobs/job/job_webhook.go]: done

File: pkg/controller/jobs/job/job_webhook.go [hunk 2]
```diff
@@ -117,6 +124,20 @@ func validatePartialAdmissionCreate(job *Job) field.ErrorList {
 			}
 		}
 	}
+	if strVal, found := job.Annotations[JobSyncCompletionWithParallelism]; found {
+		enabled, err := strconv.ParseBool(strVal)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(syncCompletionAnnotationsPath, job.Annotations[JobSyncCompletionWithParallelism], err.Error()))
+		}
+		if enabled {
+			if job.Spec.CompletionMode == nil || *job.Spec.CompletionMode == batchv1.NonIndexedCompletion {
```
[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/controller/jobs/job/job_webhook.go:133]: This is my misunderstanding.
If the `job.Spec.CompletionMode` is nil, `*job.Spec.CompletionMode == batchv1.NonIndexedCompletion` will not be performed.

**Review by tenzen-y — COMMENTED** [review]
/lgtm
/assign @alculquicondor

File: pkg/controller/jobs/job/job_controller.go [hunk 3]
```diff
@@ -51,7 +51,8 @@ var (
 )
 
 const (
-	JobMinParallelismAnnotation = "kueue.x-k8s.io/job-min-parallelism"
+	JobMinParallelismAnnotation      = "kueue.x-k8s.io/job-min-parallelism"
+	JobSyncCompletionWithParallelism = "kueue.x-k8s.io/job-completions-equal-parallelism"
```
Comment (alculquicondor) [inline pkg/controller/jobs/job/job_controller.go:55]:
```suggestion
	JobCompletionsEqualParallelismAnnotation = "kueue.x-k8s.io/job-completions-equal-parallelism"
```

Comment (trasc) [reply pkg/controller/jobs/job/job_controller.go]: done

File: pkg/controller/jobs/job/job_webhook.go [hunk 4]
```diff
@@ -81,6 +83,11 @@ func (w *JobWebhook) Default(ctx context.Context, obj runtime.Object) error {
 	}
 
 	jobframework.ApplyDefaultForSuspend(job, w.manageJobsWithoutQueueName)
+
+	if job.Spec.Completions == nil && job.syncCompletionWithParallelism() {
```
[the remark this answers is not in the document]
Comment (trasc) [reply pkg/controller/jobs/job/job_webhook.go]: done

File: pkg/controller/jobs/job/job_webhook.go [hunk 5]
```diff
@@ -117,6 +124,20 @@ func validatePartialAdmissionCreate(job *Job) field.ErrorList {
...
+		enabled, err := strconv.ParseBool(strVal)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(syncCompletionAnnotationsPath, job.Annotations[JobSyncCompletionWithParallelism], err.Error()))
+		}
+		if enabled {
+			if job.Spec.CompletionMode == nil || *job.Spec.CompletionMode == batchv1.NonIndexedCompletion {
+				allErrs = append(allErrs, field.Invalid(syncCompletionAnnotationsPath, job.Annotations[JobSyncCompletionWithParallelism], "should not be enabled for NonIndexed jobs"))
+			}
+			if pointer.Int32Deref(job.Spec.Parallelism, 1) != pointer.Int32Deref(job.Spec.Completions, 1) {
+				allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "completions"), job.Spec.Completions, "should be equal to parallelism when sync completions is enabled"))
```
[the remark this answers is not in the document]
Comment (trasc) [reply pkg/controller/jobs/job/job_webhook.go]: done

[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/controller/jobs/job/job_webhook.go]: > But I'm open to suggestions for better names for the annotation.

I'm ok with `kueue.x-k8s.io/job-completions-equal-parallelism`.

**Comment by alculquicondor** [comment]
/release-note-edit
```release-note
An option to synchronize batch/job.completions with parallelism in case of partial admission
```

**Pull request merged**